File: RxUI/src/main/java/com/tamsiree/rxui/activity/ActivityWebView.java
package com.tamsiree.rxui.activity;


import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.webkit.DownloadListener;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.tamsiree.rxtool.RxBarTool;
import com.tamsiree.rxtool.RxConstants;
import com.tamsiree.rxtool.RxImageTool;
import com.tamsiree.rxtool.RxKeyboardTool;
import com.tamsiree.rxui.R;
import com.tamsiree.rxui.view.RxTextAutoZoom;

/**
 * @author tamsiree
 */
public class ActivityWebView extends ActivityBase {

    /**
     * milliseconds, desired time passed between two back presses.
     */
    private static final int TIME_INTERVAL = 2000;
    ProgressBar pbWebBase;
    TextView tvTitle;
    WebView webBase;
    ImageView ivFinish;
    RxTextAutoZoom mRxTextAutoZoom;
    LinearLayout llIncludeTitle;
    private String webPath = "";
    private long mBackPressed;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        RxBarTool.setTransparentStatusBar(this);
        setContentView(R.layout.activity_webview);
        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        initView();// 初始化控件 - FindViewById之类的操作
        initData();// 初始化控件的数据及监听事件
    }

    private void initView() {
        // TODO Auto-generated method stub
        mRxTextAutoZoom = findViewById(R.id.afet_tv_title);
        llIncludeTitle = findViewById(R.id.ll_include_title);
        tvTitle = findViewById(R.id.tv_title);
        pbWebBase = findViewById(R.id.pb_web_base);
        webBase = findViewById(R.id.web_base);
        ivFinish = findViewById(R.id.iv_finish);
        ivFinish.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (webBase.canGoBack()) {
                    webBase.goBack();
                } else {
                    finish();
                }
            }
        });

        initAutoFitEditText();
    }

    public void initAutoFitEditText() {

        mRxTextAutoZoom.clearFocus();
        mRxTextAutoZoom.setEnabled(false);
        mRxTextAutoZoom.setFocusableInTouchMode(false);
        mRxTextAutoZoom.setFocusable(false);
        mRxTextAutoZoom.setEnableSizeCache(false);
        //might cause crash on some devices
        mRxTextAutoZoom.setMovementMethod(null);
        // can be added after layout inflation;
        mRxTextAutoZoom.setMaxHeight(RxImageTool.dip2px(55f));
        //don't forget to add min text size programmatically
        mRxTextAutoZoom.setMinTextSize(37f);

        RxTextAutoZoom.setNormalization(this, llIncludeTitle, mRxTextAutoZoom);

        RxKeyboardTool.hideSoftInput(this);
    }

    @SuppressLint({"SetJavaScriptEnabled", "ObsoleteSdkInt"})
    private void initData() {
        //设置加载进度最大值
        pbWebBase.setMax(100);

//        webPath = getIntent().getStringExtra("URL");

        //加载的URL
        webPath = RxConstants.URL_BAIDU_SEARCH;
        if (webPath.equals("")) {
            webPath = "http://www.baidu.com";
        }
        WebSettings webSettings = webBase.getSettings();
        if (Build.VERSION.SDK_INT >= 19) {
            //加载缓存否则网络
            webSettings.setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);
        }

        if (Build.VERSION.SDK_INT >= 19) {
            //图片自动缩放 打开
            webSettings.setLoadsImagesAutomatically(true);
        } else {
            //图片自动缩放 关闭
            webSettings.setLoadsImagesAutomatically(false);
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            //软件解码
            webBase.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        }
        //硬件解码
        webBase.setLayerType(View.LAYER_TYPE_HARDWARE, null);

//        webSettings.setAllowContentAccess(true);
//        webSettings.setAllowFileAccessFromFileURLs(true);
//        webSettings.setAppCacheEnabled(true);
   /*     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webSettings.setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
        }*/


        // setMediaPlaybackRequiresUserGesture(boolean require) //是否需要用户手势来播放Media，默认true

        // 设置支持javascript脚本
        webSettings.setJavaScriptEnabled(true);

//        webSettings.setPluginState(WebSettings.PluginState.ON);

        // 设置可以支持缩放
        webSettings.setSupportZoom(true);
        // 设置出现缩放工具 是否使用WebView内置的缩放组件，由浮动在窗口上的缩放控制和手势缩放控制组成，默认false
        webSettings.setBuiltInZoomControls(true);

        //隐藏缩放工具
        webSettings.setDisplayZoomControls(false);
        // 扩大比例的缩放
        webSettings.setUseWideViewPort(true);

        //自适应屏幕
        webSettings.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.SINGLE_COLUMN);
        webSettings.setLoadWithOverviewMode(true);

        webSettings.setDatabaseEnabled(true);
        //保存密码
        webSettings.setSavePassword(true);
        //是否开启本地DOM存储  鉴于它的安全特性（任何人都能读取到它，尽管有相应的限制，将敏感数据存储在这里依然不是明智之举），Android 默认是关闭该功能的。
        webSettings.setDomStorageEnabled(true);
        webBase.setSaveEnabled(true);
        webBase.setKeepScreenOn(true);


        // 设置setWebChromeClient对象
        webBase.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onReceivedTitle(WebView view, String title) {
                super.onReceivedTitle(view, title);
                mRxTextAutoZoom.setText(title);
            }

            @Override
            public void onProgressChanged(WebView view, int newProgress) {
                // TODO Auto-generated method stub
                pbWebBase.setProgress(newProgress);
                super.onProgressChanged(view, newProgress);
            }
        });
        //设置此方法可在WebView中打开链接，反之用浏览器打开
        webBase.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                if (!webBase.getSettings().getLoadsImagesAutomatically()) {
                    webBase.getSettings().setLoadsImagesAutomatically(true);
                }
                pbWebBase.setVisibility(View.GONE);
            }

            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                // TODO Auto-generated method stub
                pbWebBase.setVisibility(View.VISIBLE);
                super.onPageStarted(view, url, favicon);
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {

                if (url.startsWith("http:") || url.startsWith("https:")) {
                    view.loadUrl(url);
                    return false;
                }


                // Otherwise allow the OS to handle things like tel, mailto, etc.
                try {
                    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
                    startActivity(intent);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return true;
            }
        });
        webBase.setDownloadListener(new DownloadListener() {
            @Override
            public void onDownloadStart(String paramAnonymousString1, String paramAnonymousString2, String paramAnonymousString3, String paramAnonymousString4, long paramAnonymousLong) {
                Intent intent = new Intent();
                intent.setAction("android.intent.action.VIEW");
                intent.setData(Uri.parse(paramAnonymousString1));
                startActivity(intent);
            }
        });

        webBase.loadUrl(webPath);
        Log.v("帮助类完整连接", webPath);
//        webBase.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.FILL_PARENT,webBase.getHeight()));
    }

    public void setWebPath(String url) {
        webPath = url;
        webBase.loadUrl(webPath);
        Log.v("设置新的URL：", webPath);
    }

    @Override
    protected void onSaveInstanceState(Bundle paramBundle) {
        super.onSaveInstanceState(paramBundle);
        paramBundle.putString("url", webBase.getUrl());
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        try {
            super.onConfigurationChanged(newConfig);
            if (mContext.getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
                Log.v("Himi", "onConfigurationChanged_ORIENTATION_LANDSCAPE");
            } else if (mContext.getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT) {
                Log.v("Himi", "onConfigurationChanged_ORIENTATION_PORTRAIT");
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void onBackPressed() {

        if (webBase.canGoBack()) {
            webBase.goBack();
        } else {
            if (mBackPressed + TIME_INTERVAL > System.currentTimeMillis()) {
                super.onBackPressed();
                return;
            } else {
                Toast.makeText(getBaseContext(), "再次点击返回键退出", Toast.LENGTH_SHORT).show();
            }
            mBackPressed = System.currentTimeMillis();
        }
    }

}

File: RxUI/src/main/java/com/tamsiree/rxui/activity/ActivityWebView.kt
package com.tamsiree.rxui.activity

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import com.tamsiree.rxkit.RxBarTool.setTransparentStatusBar
import com.tamsiree.rxkit.RxConstants
import com.tamsiree.rxkit.RxWebViewTool.initWebView
import com.tamsiree.rxkit.TLog
import com.tamsiree.rxkit.interfaces.OnWebViewLoad
import com.tamsiree.rxkit.view.RxToast
import com.tamsiree.rxui.R
import kotlinx.android.synthetic.main.activity_webview.*
import kotlinx.android.synthetic.main.include_webview.*

/**
 * @author tamsiree
 */
class ActivityWebView : ActivityBase() {

    private var webPath = ""
    private var mBackPressed: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTransparentStatusBar(this)
        setContentView(R.layout.activity_webview)
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN)

    }

    // 初始化控件 - FindViewById之类的操作
    override fun initView() {
        web_rx_title.setLeftOnClickListener(View.OnClickListener {
            if (web_base.canGoBack()) {
                web_base.goBack()
            } else {
                finish()
            }
        })
    }

    // 初始化控件的数据及监听事件
    @SuppressLint("SetJavaScriptEnabled")
    override fun initData() {
        //设置加载进度最大值
        pb_web_base.max = 100

//        webPath = getIntent().getStringExtra("URL");

        //加载的URL
        webPath = RxConstants.URL_BAIDU_SEARCH
        if (webPath == "") {
            webPath = "http://www.baidu.com"
        }
        initWebView(mContext, web_base, object : OnWebViewLoad {
            override fun onPageStarted() {
                pb_web_base.visibility = View.VISIBLE
            }

            override fun onReceivedTitle(title: String) {
                web_rx_title.setTitle(title)
            }

            override fun onProgressChanged(newProgress: Int) {
                pb_web_base.progress = newProgress
            }

            override fun shouldOverrideUrlLoading() {}
            override fun onPageFinished() {
                pb_web_base.visibility = View.GONE
            }
        })
        web_base.loadUrl(webPath)
        TLog.v("帮助类完整连接", webPath)
        //        webBase.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.FILL_PARENT,webBase.getHeight()));
    }

    fun setWebPath(url: String) {
        webPath = url
        web_base.loadUrl(webPath)
        TLog.v("设置新的URL：", webPath)
    }

    override fun onSaveInstanceState(paramBundle: Bundle) {
        super.onSaveInstanceState(paramBundle)
        paramBundle.putString("url", web_base.url)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        try {
            super.onConfigurationChanged(newConfig)
            if (mContext.resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
                TLog.v("Himi", "onConfigurationChanged_ORIENTATION_LANDSCAPE")
            } else if (mContext.resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) {
                TLog.v("Himi", "onConfigurationChanged_ORIENTATION_PORTRAIT")
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    override fun onBackPressed() {
        if (web_base.canGoBack()) {
            web_base.goBack()
        } else {
            if (mBackPressed + TIME_INTERVAL > System.currentTimeMillis()) {
                super.onBackPressed()
                return
            } else {
//                Toast.makeText(getBaseContext(), "再次点击返回键退出", Toast.LENGTH_SHORT).show();
                RxToast.normal("再次点击返回键退出")
            }
            mBackPressed = System.currentTimeMillis()
        }
    }

    companion object {
        /**
         * milliseconds, desired time passed between two back presses.
         */
        private const val TIME_INTERVAL = 2000
    }
}